oeps/oep-0004.rst

```diff
@@ -0,0 +1,177 @@
+
+Rationale
+=========
+
+The discussion of scopes and application authorization has come up many times, but only recently have we (at edx.org) started to see a meaningful number of requests from external application developers that want to build on our platform.
+
+For example, certain applications would like access to our edx.org course catalog for use cases like affiliate marketing and content repurposing. In this case, we want to be able to grant applications read-only access to our catalog, but not write access, and not access to any other resources (such as enrollments or user profiles).
+
+Specification
+=============
+
+Data model
+----------
+
+============ ==================================================================
+Object       Description
+============ ==================================================================
+service user an Open edX account created solely for server-to-server
+             integrations; can have multiple associated applications
+application  an autonomous system that needs to integrate with an Open edX
+             platform, owned by a single service user
+resource     named grouping of related capabilities. Will generally (not
+             always) be a noun like "catalog"
+action       describes a high-level behavior and will generally be "read" or
+             "write", but resources may use non-standard definitions where
+             absolutely necessary (should be the exception)
+scope        a single authorization (naming convention: ``resource:action``)
+             for an application relative to an Open edX instance; should only
+             describe broad application authorization (as opposed to, say,
+             fine-grained user authorizations)
+access token a temporary credential that grants a set of scopes to an
+             application
+============ ==================================================================
+
+I sketched these relationships out in the diagram below. Note that instead of using "read/write" as actions, I noted the HTTP verbs instead, just to give a sense for the implementation details we've got in mind at this point. Same ideas though.
+
+.. image:: oep-0004/authz_relationships.jpg
+  :alt: A sketch of the proposed data model that connects service users to
+        resources through applications, scopes, and resource actions.
+
+Governance
+----------
+* Centralize the complete list of scopes but decentralize the associated
+  metadata and definitions
+
+* Centralize application and token management but decentralize token
+  enforcement
+
+* Centralize common implementation patterns as we encounter them (logging,
+  enforcement, etc)
+
+* Access tokens must be reasonably well-contained `JSON Web Tokens`_, signed by
+  the central authorization service, that contain scope information
```

Comment:
@efagin: +1 for removing the JWT reference and moving to the more generic language. I think that the important thing to get across is that the access tokens will be functionally opaque to the clients, and we can separately discuss whether the services accepting that token need to do some lookup to an auth service, interpret the JWT, etc.